primitives/state-machine/src/overlayed_changes.rs

```diff
@@ -149,10 +149,17 @@ impl OverlayedChanges {
 	/// Inserts the given key-value pair into the prospective child change set.
 	///
 	/// `None` can be used to delete a value specified by the given key.
-	pub(crate) fn set_child_storage(&mut self, storage_key: Vec<u8>, key: Vec<u8>, val: Option<Vec<u8>>) {
+	pub(crate) fn set_child_storage(
+		&mut self,
+		storage_key: Vec<u8>,
+		child_info: ChildInfo,
+		key: Vec<u8>,
+		val: Option<Vec<u8>>,
+	) {
 		let extrinsic_index = self.extrinsic_index();
-		let map_entry = self.prospective.children.entry(storage_key).or_default();
-		let entry = map_entry.entry(key).or_default();
+		let map_entry = self.prospective.children.entry(storage_key)
+			.or_insert_with(|| (Default::default(), child_info.to_owned()));
+		let entry = map_entry.0.entry(key).or_default();
```

Comment:
I just realize that this can go wrong if a user change child info and spawn a child info with root or a child_info without root (then Eq will mismatch). But I think it would indicate bad use of the api (if you memoize root you should do it globabally and not use some new child_info everywhere.
I can also change Eq to only make root being None always equal to root being Some.

Reply:
> if user change child info

what does that mean ? like in the future when there is another extension one could change the extension here ? would it work ?
I thought changing the extension wasn't doable.

Reply:
By user I meant runtime developer, but it would be a bug. (triggered by using a child trie api both with root memoization and without).